### PR TITLE
FastSim: new rechits containers

### PR DESCRIPTION
**Pull request by lveldere**
- fastsim tracker rechit classes are replaced with
  - a fastsim base class FastTrackerRecHit
  - a fastsim class for single rechits FastSingleTrackerRecHit
    (accomodates the fastsim equivalents of SiPixelRecHit, SiStripRecHit2D and SiStripRecHit1D)
  - a fastsim equivalent of SiStripMatchedRecHit2D, FastMatchedTrackerRecHit
  - a fastsim equivalent of ProjectedSiStripRecHit2D, FastProjectedTrackerRecHit
- trackerHitRTTI::RTTI is extended to accommodate the new rechit containers
- a new implementation of FastTrackerRecHit::sharesInput
  ( will receive further study once merged)
- rechit production is further modularised
  - actual rechit production by the old SiTrackerGaussianSmearingRecHitConverter
  - hit matching by new FastTrackerRecHitMatcher
  - predefined combinations of hits by new FastTrackerRecHitCombiner
- SiTrackerGaussianSmearingRecHitConverter produces now
  OwnVector<FastTrackerRecHit>
  vector<\Ref<FastTrackerRecHitCollection> > with one entry per SimHit to find for each SimHit the corresponding RecHit
- predefined combinations of hits are now in a new, simple format:
  vector<vector<Ref<FastTrackerRecHitCollection> > >
- model for hit splitting has been improved and moved to a separate class FastTrackerRecHitSplitter
- model for masking of hits in iterative tracking improved
  (see HitMaskHelper and FastTrackerRecHitMaskProducer)
- GlobalPixelTracking is not fully migrated yet (it produces zero result):
  it was not properly configured to begin with,
  and we need to study the fullsim equivalent before fixing it
- various code files were updated for the new rechit classes and the new rechit combination format
- y position of strip hits is set to 0 (before it was the simhit y-position)

### CODE REVIEW

**Comment by lveldere** [comment]
please test

File: DataFormats/TrackerRecHit2D/src/classes_def.xml [hunk 1]
```diff
@@ -163,4 +124,37 @@
   <class name="reco::ClusterRemovalInfo::SiPixelClusterRefProd" />
   <class name="reco::ClusterRemovalInfo::SiStripClusterRefProd" />
   <class name="edm::Wrapper<reco::ClusterRemovalInfo>" />
+
+  // fastsim containers
+  
+  <class name="FastTrackerRecHit" ClassVersion="10">
```
Comment (slava77) [inline DataFormats/TrackerRecHit2D/src/classes_def.xml]: do we care about any old(er) fastsim files with these hits stored? (or are the hits usually not kept?)

File: FastSimulation/Tracking/interface/FastTrackingHelper.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+#include "DataFormats/TrackerRecHit2D/interface/FastTrackerRecHitCollection.h"
+#include "DataFormats/TrackerRecHit2D/interface/FastTrackerRecHit.h"
+
+namespace fastTrackingHelper {
+    
+    template<class T> inline void setRecHitCombinationIndex(edm::OwnVector<T> & recHits,int32_t icomb){
+	for(auto & recHit : recHits){
+	    if(!trackerHitRTTI::isFast(recHit)){
+		edm::LogError("fastTrackingHelpers::setRecHitCombinationIndex") << " one of hits in OwnVector is non-fastsim" << std::endl;
+		exit(1);
```
Comment (slava77) [inline FastSimulation/Tracking/interface/FastTrackingHelper.h]: throw a cms exception; 
direct job control system calls are not allowed

File: DataFormats/TrackerRecHit2D/src/classes_def.xml (hunk 1 above)
Comment (lveldere) [inline DataFormats/TrackerRecHit2D/src/classes_def.xml]: hi slava

I would be very surprised if there were any interest in running with 76 on samples with older releases. But I could put the old classes_def lines back if that is desirable and technically possible.
(I have no clue how CMSSW will react having lines for classes that no longer exist)

Comment (slava77) [inline DataFormats/TrackerRecHit2D/src/classes_def.xml]: I had in mind iorules or converters if need to read old info.

**Comment by lveldere** [comment]
I have completed the description of the pull request.
See top of the conversation.

**Comment by slava77** [comment]
the errors show up during reading of old files

**Comment by slava77** [comment]
About ``` a new, simple format: vector<vector<Ref<FastTrackerRecHitCollection> > > ```
we have advised against using vectors of vectors for data storage.
In this case we will have fairly large blocks.
An alternative solution is to linearize the vector of vectors and keep track of the second dimension with a vector of begin/end indices.
It's unclear though if it has to be done this way.

@VinInn @Dr15Jones any comment?

**Comment by Dr15Jones** [comment]
I agree that we should avoid storing ```vector<vector<...>>```.

**Comment by matt-komm** [comment]
what about an edm::RangeMap instead of the vector<vector<...>> to store the combinations? std::vector<edm::Ref<>> => edm::RefVector<>?

**Comment by VinInn** [comment]
in memory and with C++11 vector<vector<>> is not anymore so evil if each inner vector properly sized/reserved before filling
one can try to use edmNew::DetSetVector that replaced RangeMap a decade ago

**Comment by lveldere** [comment]
I replaced exit with cms exception.
I applied shrink_to_fit on inner vectors to improve memory behaviour.

@Dr15Jones @slava77 @VinInn @matt-komm 
The combinations are not supposed to be stored in output files.
(right now, I don't see a case for that)
With the shrink_to_fit applied and given that the combinations will not be stored,
and given that those collections are actually rather light (just vector<vector<Ref> > ),
is there much reason to change?

In general, if any refactoring is required, on this topic or other ones,
I would prefer to deal with that in next pull requests,
because this one got a bit out of hand, and is a pain to work with.

**Comment by lveldere** [comment]
@slava77 

Is there a way to get the comparisons for the succeeded tests?
I imagine the absence of those comparisons will obstruct the approval of this pr...

**Comment by slava77** [comment]
On 8/28/15 3:09 AM, lveldere wrote:
> @slava77 <https://github.com/slava77>
> 
> Is there a way to get the comparisons for the succeeded tests?
> I imagine the absence of those comparisons will obstruct the approval of
> this pr...
> 

Comparisons are there for the matrix workflows.
The failures were in the unit tests.

We can keep the old classes to have dictionaries to read those old files
until a new pre-release shows up, then we can update the unit tests to
read these new files
and drop the old class files.

**Comment by lveldere** [comment]
Thanks Slava.
Could you tell me how I run the problematic unit tests.
Lukas

**Comment by Dr15Jones** [comment]
Since this type is not intended to be stored in any data tier then I'd vote for only changing if it is shown to be a memory/speed problem. [std::vector<std::vector<..>> do not have good CPU cache behaviors]

**Comment by slava77** [comment]
about rerunning the tests:
to pick out the specific failing test
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10987/7697/unitTests.log
find the "ERRORS" e.g. for testRecoMETMETProducers
then look for Test "testRecoMETMETProducers" .. it will tell you that RecoMET/METProducers/test/runtests.sh is running and from the log you can see that the second test failed
this is
RecoMET/METProducers/test/recoMET_caloMet_cfg.py

if you just want to run all tests blindly,
then in your test area "scram b runtests "
(since you probably have a half of the release in dependencies, manual cherry-picking of the test script may be more appropriate)

**Comment by lveldere** [comment]
@slava77 

It's not that easy to restore this classes_def.xml.
In my area CMSSW seems to have forgotten about the old classes,
and that leads to compilation errors when I put back lines in classes_def.xml.
Restoring everything will cost me a lot of time...

### Alternative 1

I edit the affected unit tests to produce the fastsim sample on the fly.
Once there is a proper relval we revert to the old way.

### Alternative 2

We live with a few failing unit tests.

I'm able to let 5/6 tests in  RecoMET/METProducers/test/runtests.sh
run properly by changing the input to fullsim.
Number 6 depends on a collection that is only stored in FastSim (pfclusters).

The failing pat unit test is fastsim specific,
and I personally could live w/o it, until there are new RelVal samples.

**Comment by civanch** [comment]
Hi Lukas,

I think alternative2 is not a solution: having any unit test failing practically may shadow new problems in new PRs. Ideally unit test should test something useful, if it is absolete it should be removed, if not removed then should not fail.

**Comment by slava77** [comment]
Lukas,

Please check if that test will run by dropping the hits upfront in the
input drop statements.
If yes, this should be just added to the test configs
 (since the test is for MET and not for tracking hits).

**Comment by lveldere** [comment]
@slava77 

Dropping the hits from the input did not work.
Somehow, many collections (more than I expected) depend on the hit collection.
Probably there are indirect dependencies.
All dependent collections are removed, some of which are required by the unit tests.

So, I produce the required fastsim samples on the fly...

**Comment by lveldere** [comment]
please test

**Comment by lveldere** [comment]
please test

**Comment by lveldere** [comment]
needed rebase,
replaced by #11078

**Pull request closed without merging**